src/Features/Core/Portable/ConvertIfToSwitch/AbstractConvertIfToSwitchCodeRefactoringProvider.Analyzer.cs

```diff
@@ -0,0 +1,397 @@
+                    // i >= 5
+                    GreaterThanOrEqual when IsConstant(op.RightOperand) => (BoundKind.Lower, op.LeftOperand, op.RightOperand),
+                    _ => default
+                };
+            }
+
+            private static BinaryOperatorKind Negate(BinaryOperatorKind operatorKind)
+            {
+                return operatorKind switch
+                {
+                    LessThan => GreaterThan,
+                    LessThanOrEqual => GreaterThanOrEqual,
+                    GreaterThanOrEqual => LessThanOrEqual,
+                    GreaterThan => LessThan,
+                    NotEquals => NotEquals,
+                    var v => throw ExceptionUtilities.UnexpectedValue(v)
+                };
+            }
+
+            private static bool IsComparisonOperator(BinaryOperatorKind operatorKind)
+            {
+                switch (operatorKind)
+                {
+                    case LessThan:
+                    case LessThanOrEqual:
+                    case GreaterThanOrEqual:
+                    case GreaterThan:
+                    case NotEquals:
+                        return true;
+                    default:
+                        return false;
+                }
+            }
+
+            private static bool IsConstant(IOperation operation)
+            {
+                // Use syntax to skip possible conversion operation
+                return operation.SemanticModel.GetConstantValue(operation.Syntax).HasValue;
+            }
+
+            private bool CheckTargetExpression(IOperation operation)
+            {
+                if (operation is IConversionOperation { IsImplicit: false } op)
+                {
+                    // Unwrap explicit casts because switch will emit those anyways
+                    operation = op.Operand;
+                }
+
+                var expression = operation.Syntax;
+                // If we have not figured the switch expression yet,
+                // we will assume that the first expression is the one.
+                if (_switchTargetExpression is null)
+                {
+                    _switchTargetExpression = expression;
+                    return true;
+                }
+
+                return _syntaxFacts.AreEquivalent(expression, _switchTargetExpression);
+            }
+
+            public abstract bool CanConvert(IConditionalOperation operation);
+            public abstract bool HasUnreachableEndPoint(IOperation operation);
+
+            public abstract bool SupportsSwitchExpression { get; }
+            public abstract bool SupportsRelationalPattern { get; }
+            public abstract bool SupportsSourcePattern { get; }
+            public abstract bool SupportsRangePattern { get; }
+            public abstract bool SupportsTypePattern { get; }
+            public abstract bool SupportsCaseGuard { get; }
```

Comment:
nit: move to type.

Reply:
these are being used in the Analyzer itself and it does not have access to the parent type

Reply:
ah you probably meant *top*

Reply:
yes :)